### ISSUE TITLE
Add support for subscriptions

### ISSUE THREAD

**Issue by PeterStaev** [PeterStaev/nativescript-purchase#2]

**Comment by ignaciolarranaga** [comment]
What do you plan to do here Peter ?, I'm already using your plugin for subscriptions.
Basically I understood that as I add the listener in the app startup, every time a subscription gets renewed the even will follow the normal flow, so voila.

Is that correct ?, or I'm missing something ?

**Comment by PeterStaev** [comment]
Hey @ignaciolarranaga , I was not sure that it will seamlessly work on iOS. For Android will definitely need separate implementation.

**Comment by ignaciolarranaga** [comment]
Thanks @PeterStaev, did not get into the Android implementation yet. Any further feedback (if you think it won't work) will be appreciated!

**Comment by PeterStaev** [comment]
@ignaciolarranaga , Android support for the plugin is still in the works. I have not researched subscriptions yet.

**Comment by danieljohngomez** [comment]
would love to have this feature

**Comment by nnieto** [comment]
@ignaciolarranaga it works for you the subscriptions? I always get the transaction state failed but in appstore appears subscribed

**Comment by erjdriver** [comment]
Does this plug-in work for iOS renewable subscriptions?

If so, where does shared secret id come in.

Thanks.

**Comment by PeterStaev** [comment]
@erjdriver , like the title of the issues says, subscription support is not yet implemented 😄

**Comment by erjdriver** [comment]
Wow that was quick response!

Ok wasn't sure because someone wrote above that he was already using the plug-in for subscriptions.

I know zero about iOS development and this feature would take years for me to add (in {N}).

Without being a PITA, is there an approximate timeline. Donations..patreon...to get this added?

Thank you for your time.

**Comment by davecoffin** [comment]
@erjdriver ...I am using this plugin for iOS renewable subscriptions. 
```
exports.subscribe = function() {
    purchase.getProducts().then(function(products) {
        if (!products.length) {
            alert('Sorry, there was an error subscribing you. Please contact support.');
            return;
        }
        parentsMonthly = products[0];

        purchase.on(purchase.transactionUpdatedEvent, function(transaction) {
            if (transaction.transactionState === transactionModule.TransactionState.Purchasing) {
                //console.log('we are purchasing!');
            } else if (transaction.transactionState === transactionModule.TransactionState.Purchased) {
                successfullyPaid();
            }
            else if (transaction.transactionState === transactionModule.TransactionState.Restored) {
                successfullyPaid();
            }
            else if (transaction.transactionState === transactionModule.TransactionState.Failed) {
                dialogsModule.alert({
                    title: "Hmm...",
                    message: "The transaction failed or you canceled the transaction.",
                    okButtonText: "Thanks!"
                }).then(function () {
                    successfullyPaid('failed_app_transaction');
                });
            } else {
                successfullyPaid('unknown_app_transaction');
            }
        });

        if (purchase.canMakePayments()) {
            purchase.buyProduct(parentsMonthly);
        } else {
            alert("Sorry, your account is not eligible to make payments!");
        }
    });
}```

**Comment by erjdriver** [comment]
Is this a function you wrote?

I believe I'm doing something similar and seems to be working. But not sure whether I'm getting the subscription "rate" or not. App is not fully developed.

On the Apple connect site, I saw something about a shared-secret . Dont need that?

This is what the site says:

"The shared secret is a unique code to receive receipts for all your auto-renewable in-app purchases. Without the shared secret, you won’t be able to test or offer auto-renewable subscriptions."

**Comment by ignaciolarranaga** [comment]
@erjdriver On the renewal you need to retrieve the receipt from apple (because it tells you for example how much time the subscription last). To make this request you need the secret.

**Comment by ignaciolarranaga** [comment]
@NNieto I did not release the product with the subscriptions (jump into another project). They worked fine on sandbox but can't tell you on prod.

**Comment by erjdriver** [comment]
@ignaciolarranaga - can you tell me how I can get this plugin to work with renewable subscriptions - especially on the ios side.

**Comment by ignaciolarranaga** [comment]
For Android there is no implementation yet, for iOS I can share the code I was using which is basically similar to the @davecoffin one:

    private initializeAppPurchases() {
        purchase.init(["orders", "images", "employees"]);

        // Process postponed transactions
        this._purchaseService.reprocessPostponedTransactions();

        // Registering the handler for transaction updates
        purchase.on(purchase.transactionUpdatedEvent, (transaction: Transaction) => {
            console.log("Processing " + transaction.transactionState +
                " transaction for " + transaction.productIdentifier +
                (transaction.transactionDate ? " of " + transaction.transactionDate : ""));
            
            if (transaction.transactionState == TransactionState.Purchased) {
                // Processing payment
                let purchaseTransaction = new PurchaseTransaction();
                purchaseTransaction.storeId = transaction.transactionIdentifier;
                purchaseTransaction.date = transaction.transactionDate;
                purchaseTransaction.receipt = transaction.transactionReceipt;
                purchaseTransaction.state = PurchaseTransaction.STATE_PURCHASED;

                // Mapping product identifiers to Features
                switch (transaction.productIdentifier) {
                    case AppConfig.PRODUCT_IMAGES: purchaseTransaction.feature = PurchaseTransaction.FEATURE_IMAGES; break;
                    case AppConfig.PRODUCT_EMPLOYEES: purchaseTransaction.feature = PurchaseTransaction.FEATURE_EMPLOYEES; break;
                }

                this._purchaseService.processTransaction(purchaseTransaction);
            }
            // We do nothing with: TransactionState.Purchasing (the initial status), TransactionState.Failed, TransactionState.Restored & TransactionState.Deferred
        });
    }

Basically every time the app starts you initialize the plugin, it will receive the transactions for processing and depending on you choice you will process them on the app or the backend.
In my case it was on the backend so I added a service for that `(this._purchaseService.processTransaction(purchaseTransaction);`, so you just send the information received and from the backend (specially the encoded receipt) you call the apple service to get the decoded receipt (as I mentioned you normally need the expiration of the subscription to update your DB of active subscriptions which is inside the encoded receipt).

That's basically all.

As I mentioned I never ran it on production, just sandbox (the sandbox expires the subscriptions every few minutes so you can test the renewals).

Good luck, it is a little tricky but not so much.

One minor advice. Have in mind (because I had to make some particular adjustments on this) that you might receive many transactions almost in parallel, so you code/backend has to be able to execute them without deadlocks (i.e. take care of the transaction isolation). It might sound an uncommon scenario but if you open the application after several renewals happen, you will receive all of them at once, like extending the subscription for a month, another month, another month, and so on). And obviously take care of keep the latest renewal according to the receipt, not the last received.

**Comment by erjdriver** [comment]
@ignaciolarranaga - thank you.

Any way I can get you to share the purchase-service class???

**Comment by ignaciolarranaga** [comment]
It is kind of specific and also java backend (not sure what you are using), but here it is:

```
import com.google.gson.Gson;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import com.gynto.services.PurchaseServices;
import java.io.IOException;
import org.apache.http.HttpResponse;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClientBuilder;
import org.apache.http.util.EntityUtils;
import org.json.simple.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import com.gynto.model.repositories.PurchaseTransactionRepository;
import com.gynto.services.exceptions.UnableToDetermineCompanyForPurchaseException;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Set;
import javax.persistence.EntityManager;
import javax.persistence.LockModeType;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.TransactionDefinition;
import org.springframework.transaction.TransactionStatus;
import org.springframework.transaction.support.TransactionCallbackWithoutResult;
import org.springframework.transaction.support.TransactionTemplate;

@Service
public class PurchaseServicesImpl implements PurchaseServices {

	private static final Logger log = LoggerFactory.getLogger(PurchaseServicesImpl.class);

	protected static class ReceiptDTO {
		public String transaction_id;				// "1000000229860434"
		public String unique_identifier;			// "85667412e0cea5c4705195b814e2b2e4aadf17bf"
		public Long purchase_date_ms;				// "1471342049000"
		public String purchase_date;				// "2016-08-16 10:07:29 Etc/GMT"
		public String purchase_date_pst;			// "2016-08-16 03:07:29 America/Los_Angeles"
		public Long expires_date;					// "1471342349000"
		public String expires_date_formatted;		// "2016-08-16 10:12:29 Etc/GMT"
		public String expires_date_formatted_pst;	// "2016-08-16 03:12:29 America/Los_Angeles"
		public String unique_vendor_identifier;		// "1F2CE1A4-3A03-4037-9BA4-875C05442A1B"

		public String bid;							// "com..."
		public String product_id;					// "orders"
		public String item_id;						// "1143964169"
		public String web_order_line_item_id;		// "1000000033059433"
		public String bvrs;							// "1.0"
		public Integer quantity;					// "1"

		public String original_transaction_id;		// "1000000229860249"
		public Long original_purchase_date_ms;		// "1471342050000"
		public String original_purchase_date;		// "2016-08-16 10:07:30 Etc/GMT"
		public String original_purchase_date_pst;	// "2016-08-16 03:07:30 America/Los_Angeles"
	}

	protected static class ValidationResponseDTO {
		public ReceiptDTO receipt;
		public ReceiptDTO latest_expired_receipt_info;

		/**
		 * https://developer.apple.com/library/ios/releasenotes/General/ValidateAppStoreReceipt/Chapters/ValidateRemotely.html#//apple_ref/doc/uid/TP40010573-CH104-SW4
		 * 21000 The App Store could not read the JSON object you provided.
		 * 21002 The data in the receipt-data property was malformed or missing.
		 * 21003 The receipt could not be authenticated.
		 * 21004 The shared secret you provided does not match the shared secret on file for your account.
		 *		Only returned for iOS 6 style transaction receipts for auto-renewable subscriptions.
		 * 21005 The receipt server is not currently available.
		 * 21006 This receipt is valid but the subscription has expired. When this status code is returned to your server, the receipt data is also decoded and returned as part of the response.
		 *		Only returned for iOS 6 style transaction receipts for auto-renewable subscriptions.
		 * 21007 This receipt is from the test environment, but it was sent to the production environment for verification. Send it to the test environment instead.
		 * 21008 This receipt is from the production environment, but it was sent to the test environment for verification. Send it to the production environment instead.
		 */
		public Integer status;
	}

	private static final Gson gson = new Gson();

	@Autowired
	private EntityManager entityManager;

	@Autowired
	private CompanyRepository companyRepository;

	@Autowired
	private PurchaseTransactionRepository purchaseTransactionRepository;

	@Autowired
    protected PlatformTransactionManager transactionManager;

	@Value("${" + com.gynto.Configuration.PROPERTY_NAME_APPLE_IN_APP_PURCHASES_VERIFY_RECEIPT_URL + "}")
	private String verifyReceiptUrl;

	@Value("${" + com.gynto.Configuration.PROPERTY_NAME_APPLE_IN_APP_PURCHASES_PASSWORD + "}")
	private String password;

	/**
	 * @param newTransaction The transaction to process
	 * @throws UnableToDetermineCompanyForPurchaseException When it is not possible to assign a company to the transaction
	 */
	@Override @PreAuthorize("isAuthenticated()")
	public void processTransaction(PurchaseTransaction newTransaction)
		throws UnableToDetermineCompanyForPurchaseException {
		try {
			// IMPORTANT: This is an slow opperation that requires external responses
			// That's why it is executed before on a separate transaction
			ValidationResponseDTO receipt = validateAppleReceipt(newTransaction.getReceipt());

			TransactionTemplate transactionTemplate = new TransactionTemplate(transactionManager);
			transactionTemplate.setIsolationLevel(TransactionDefinition.ISOLATION_SERIALIZABLE);
			transactionTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRES_NEW);
			transactionTemplate.execute(new TransactionCallbackWithoutResult() {
				@Override
				protected void doInTransactionWithoutResult(TransactionStatus status) {
					// Setting the company affected by the operation
					Company company;
					Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
					Device device = (Device) authentication.getPrincipal();
					// H2 dos not support the @Lock here becase the query has join
					Set<Company> companies = companyRepository.findAllMyPayerCompaniesByUuid(device.getUuid());
					if (companies.isEmpty()) {
						throw new RuntimeException(new UnableToDetermineCompanyForPurchaseException(
							"Unable to determine the company for the purchase. " +
							"There are no companies associated to the device.", newTransaction));
					} else if (companies.size() > 1) {
						throw new RuntimeException(new UnableToDetermineCompanyForPurchaseException(
							"Unable to determine the company for the purchase. " +
							"There is more than one company associated to the device.", newTransaction));
					} else {
						company = companies.iterator().next();
					}

					// Forcing the locks
					entityManager.lock(company, LockModeType.PESSIMISTIC_WRITE);

					// Checking if it is an update or a new transaction
					PurchaseTransaction transaction = purchaseTransactionRepository.findOneByStoreIdAndLockForWrite(newTransaction.getStoreId());
					if (transaction == null) {
						log.debug("Creating a new Transaction for " + newTransaction.getStoreId());
						transaction = newTransaction;

						// Setting the company for the transaction
						transaction.setCompany(company);

						// Indicating the transaction as not processed
						transaction.setProcessed(false);
					} else {
						log.debug("Transaction for " + newTransaction.getStoreId() + " already found. Updating....");

						// Indicating the transaction as not processed
						transaction.setProcessed(false);

						// Updating the updatable fields
						transaction.setDate(newTransaction.getDate());	// I think it might not change, but not really sure
						transaction.setReceipt(newTransaction.getReceipt());
					}

					String logMessage;
					switch (receipt.status) {
						case 21000:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The App Store could not read the JSON object you provided.";
							log.error(logMessage);
							break;
						case 21002:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The data in the receipt-data property was malformed or missing.";
							log.error(logMessage);
							break;
						case 21003:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The receipt could not be authenticated.";
							log.error(logMessage);
							break;
						case 21004:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The shared secret you provided does not match the shared secret on file for your account.";
							log.error(logMessage);
							break;
						case 21005:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The receipt server is not currently available.";
							log.error(logMessage);
							break;
						case 21007:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The receipt is from the test environment, but it was sent to the production environment for verification. " +
								"Send it to the test environment instead.";
							log.error(logMessage);
							break;
						case 21008:
							logMessage = "There was an error processing the transaction " + transaction.getStoreId() +
								": The receipt is from the production environment, but it was sent to the test environment for verification. " +
								"Send it to the production environment instead.";
							log.error(logMessage);
							break;
						case 0:
						case 21006:	// This receipt is valid but the subscription has expired. When this status code is returned to your server, the receipt data is also decoded and returned as part of the response.
							// Processing
							// "2016-08-16 03:12:29 America/Los_Angeles"
							ZonedDateTime zonedDateTime = ZonedDateTime.parse(
								receipt.receipt.expires_date_formatted_pst,
								DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss VV"));
							LocalDateTime expirationDate = zonedDateTime.withZoneSameInstant(ZoneId.systemDefault()).toLocalDateTime();

							if (company.getFeatureExpiration(transaction.getFeature()) == null) {
								company.setFeatureExpiration(transaction.getFeature(), expirationDate);
								// Due the intensive parallelism of this operations (100ms between them), and even using locking,
								// some updates where missing (i.e. when updating images expiration we lost employees, etc.),
								// so this direct update was introduced.
								//switch (transaction.getFeature()) {
								//	case IMAGES: companyRepository.setImagesFeatureExpirationByIdIfNewer(company.getId(), expirationDate); break;
								//	case EMPLOYEES: companyRepository.setEmployeesFeatureExpirationByIdIfNewer(company.getId(), expirationDate); break;
								//	default:
								//		throw new RuntimeException("Unexpected feature " + transaction.getFeature());
								//}

								logMessage = "Initializing the " + transaction.getFeature() + " service for " + company.getName() +
									" because of the transaction " + transaction.getStoreId() + ". The expiration will be set to: " + expirationDate + ".";
								log.debug(logMessage);
							} else if (expirationDate.isAfter(company.getFeatureExpiration(transaction.getFeature()))) {
								LocalDateTime oldExpirationDate = company.getFeatureExpiration(transaction.getFeature());
								company.setFeatureExpiration(transaction.getFeature(), expirationDate);
								// IMPORTANT: The isAfter is not definitive, if the operations are too close the value might be old
								// Due the intensive parallelism of this operations (100ms between them), and even using locking,
								// some updates where missing (i.e. when updating images expiration we lost employees, etc.),
								// so this direct update was introduced.
								//switch (transaction.getFeature()) {
								//	case IMAGES: companyRepository.setImagesFeatureExpirationByIdIfNewer(company.getId(), expirationDate); break;
								//	case EMPLOYEES: companyRepository.setEmployeesFeatureExpirationByIdIfNewer(company.getId(), expirationDate); break;
								//	default:
								//		throw new RuntimeException("Unexpected feature " + transaction.getFeature());
								//}

								logMessage = "Adjusting the " + transaction.getFeature() + " subscription expiration for " +
									company.getName() + " because of the transaction " + transaction.getStoreId() +
									". The expiration moves from " + oldExpirationDate + " to " + expirationDate + ".";
								log.debug(logMessage);
							} else {
								LocalDateTime currentExpirationDate = company.getFeatureExpiration(transaction.getFeature());
								logMessage = "The " + transaction.getFeature() + " expiration for " + company.getName() +
									" in the transaction " + transaction.getStoreId() + " is older than the current one, so ignoring." +
									" Current one: " + currentExpirationDate + " and the proposed one was: " + expirationDate;
								log.debug(logMessage);
							}

							// Indicating the transaction as processed
							transaction.setProcessed(true);
							break;
						default:
							logMessage = "Unrecognized status received: " + receipt.status;
							break;
					}

					transaction.setLog(logMessage);

					// Saving the transaction
					purchaseTransactionRepository.save(transaction);
				}
			});
		} catch (IOException ex) {
			// The transaction is not interrupted but the error is raised
			log.error("There was a communication error when trying to analyze the receipt on the server.", ex);
		}
	}

	@SuppressWarnings("SleepWhileInLoop")
	protected ValidationResponseDTO validateAppleReceipt(String receipt) throws IOException {
		int retries = 3;
		while (retries > 0) {
			try (CloseableHttpClient client = HttpClientBuilder.create().build()) {
				HttpPost request = new HttpPost(verifyReceiptUrl);

				JSONObject requestData = new JSONObject();
				requestData.put("receipt-data", receipt);
				requestData.put("password", password);

				StringEntity requestEntity = new StringEntity(requestData.toString());
				request.addHeader("content-type", "application/x-www-form-urlencoded");
				request.setEntity(requestEntity);
				HttpResponse response = client.execute(request);
				String strResponse = EntityUtils.toString(response.getEntity());

				return gson.fromJson(strResponse, ValidationResponseDTO.class);
			} catch (IOException ex) {
				if (retries-- <= 0) {
					log.debug("Unable to process the transaction after several retries.");
					throw ex;
				} else {
					try {
						// Waiting 1 second and retrying
						Thread.sleep(1000);
					} catch (InterruptedException ex1) {
					}
				}
			}
		}

		throw new RuntimeException("Unexpected line reached.");
	}

}

```

**Comment by erjdriver** [comment]
Ok I'm confused.

This is a Nativescript app that uses this plug-in nativescript-purchase - is calling java code in the ios version - and implements renewal subscriptions???

I see you're implementing com.gynto.services.PurchaseServices interface - where did that come from?

**Comment by ignaciolarranaga** [comment]
This is the code from my backend, that's why I mention that may be not useful. The app makes a call to the backend (REST API). The interface it just a proprietary one because of spring, just declares the method.

**Comment by erjdriver** [comment]
It's helpful - so I assume you have a NS app and it's makes a REST call to your server and the above code is running on a remote server - right?

It's helpful because I can see what needs to be done - in terms of API call to get renewal subscriptions working.

I suppose since now NS has worker threads - it could be done within the app also.

**Comment by ignaciolarranaga** [comment]
@erjdriver correct. I think you can do the full validation on NS and just call the API with the results if needed, I would say it is a personal preference.

**Comment by erjdriver** [comment]
Thanks - I appreciate your help!

**Comment by PLJean** [comment]
Has anyone figured out how to use this plugin for Android subscriptions yet? Is it just impossible in the apps current state?

If it is, I saw a line in InAppBillingHelper.java that has:

`Bundle productDetails = _service.getSkuDetails(3, _context.getPackageName(), "inapp", queryProducts);`

Wouldn't all that is needed to be done is change "inapp" to "subs" to make this a possibility? Not sure if there's more to do as I never did Android development.

**Comment by erjdriver** [comment]
Will this plug-in work on ios for non-renewable subscriptions.

If so, anything different needs to be done.


I would appreciate a response - my app is not being approved because the app needs subscriptions. :(

**Comment by nonzod** [comment]
@PeterStaev, you have any plan to include support to subscriptions for Android?

**Comment by PeterStaev** [comment]
@nonzod , like the tasks shows I have plans, but not the time right now. I'm happily accepting PRs though 😉

**Comment by nonzod** [comment]
We hope the PR #18 will help

**Comment by toddanglin** [comment]
Given the increasing shift toward subscription pricing, we should try to make this capability easier to add in {N}.

@PeterStaev What can we do from a Core Team side to help with this issue?

**Comment by PeterStaev** [comment]
Hey @toddanglin , I do not think there are any limitations in {N} that are preventing the implementation of this. It is just that the subs are more complex in terms of handling and I must find the time to deep-dive in how things are done in the native platforms so I can come up with a unified API for both platforms. I will try to find time to do this till end of the year 😄

**Comment by toddanglin** [comment]
Cool. Makes sense. Hopefully we can leverage the experience of others in this thread to short-cut the learning path for these subscription flows. @davecoffin @nonzod Let us know if you've learned anything that will help over the last 6 months.

**Comment by davecoffin** [comment]
I moved away from a subscription model, but I definitely agree core support on IAP on both platforms would be huge.

**Comment by erjdriver** [comment]
I've been using this plug-in for non-renewable IOS subscriptions.

It seems to work - but I've had 2 issues so far:
  - Don't get the transaction for some callbacks. Not sure whether it's the plug-in or my app.
  - I think apps are also supposed to handle refunds - 

Renewable subscriptions seem to be more complex because Apple handles the subscription end dates and there's this concept of "secret shared key" etc.

Even though that's better for authors (lower fees), I recommend that the project be divided in 2 steps - support the easier non-renewable subscription fully and then renewable.

Thank you very much for your effort. 

Pls set up a patreon or some other site - would be happy to contribute.

**Comment by PeterStaev** [comment]
Ok, guys, this will be soon available on NPM with version 2.0 of the plugin 🎉  Couple short notes:
* The verification and check whether a sub is currently active is left to be entirely implemented by the user's backend. This means that the plugin will not do any checks. It has methods/properties to extract the receipt data which should then be checked. This is both for security reasons (so you do not hardcode/expose the iTunesConnect shared secret) but also because the verification process is very different for Android and iOS and also between environments (sandbox or real). 
* In pre- iOS 11.2 there is absolutely no way to distinguish between in-app purchases and subscriptions. That's why the new properties `productType` and `subscriptionPeriod` will be set to `undefined` in those cases. In case you support pre- iOS 11.2 and you need to separate between subs and purchases, then you must use a different approach to separate them. What I would do is to add `.subscription.` section to my product ids and then check that one. 
* Even in iOS 11.2+ non-recurring subscriptions are indistinguishable from in-app purchases. They even work more or less the same way. So for those subs the `productType` will be set to `"inapp"`

There will a be a blog article coming soon that I will explain in details all of this and will also give examples on how to verify if a subscription is currently active or not.

**Comment by nonzod** [comment]
@toddanglin After various tests, the better way to manage autorenew subscriptions is [server to server](https://developer.apple.com/library/content/releasenotes/General/ValidateAppStoreReceipt/Chapters/ValidateRemotely.html) verification, in iOS.

**Comment by berchik** [comment]
@PeterStaev 
Thanks for the plugin.
While you're at it, subscription verification ng example would be great!

**Comment by PeterStaev** [comment]
@berchik There will be a blog article soon about subs and verification, but like explained above verification will have to be handled by your backend servers. So there is actually no difference in NG and core.

**Comment by yoat** [comment]
@PeterStaev Hi Peter, is the "blog article soon about subs and verification" available?  It would be very helpful to see it in action but I haven't found it by googling.

Thanks!

**Comment by PeterStaev** [comment]
@yoat , sadly no, and not having an ETA about it currently 😢

**Comment by adfdev** [comment]
Which one way is the best for check status of subscription?
